scarb/src/compiler/plugin/collection.rs

```diff
@@ -0,0 +1,175 @@
+use std::{collections::HashMap, sync::Arc};
+
+use anyhow::Result;
+use cairo_lang_semantic::{inline_macros::get_default_plugin_suite, plugin::PluginSuite};
+use itertools::Itertools;
+
+use crate::{
+    compiler::{CairoCompilationUnit, CompilationUnitComponentId, CompilationUnitDependency},
+    core::Workspace,
+};
+
+#[cfg(doc)]
+use crate::core::PackageId;
+
+use super::proc_macro::{ProcMacroHostPlugin, ProcMacroInstance};
+
+pub struct PluginsForComponents {
+    pub plugins: HashMap<CompilationUnitComponentId, PluginSuite>,
+    pub proc_macros: HashMap<CompilationUnitComponentId, Arc<ProcMacroHostPlugin>>,
+}
+
+impl PluginsForComponents {
+    /// Collects plugins, either built-in or procedural macros, for all components of the [`CairoCompilationUnit`].
+    pub fn collect(workspace: &Workspace<'_>, unit: &CairoCompilationUnit) -> Result<Self> {
+        let mut plugins = collect_builtin_plugins(workspace, unit)?;
+
+        let proc_macros = collect_proc_macros(workspace, unit)?
+            .into_iter()
+            .map(|(component_id, instances)| {
+                let plugin = Arc::new(ProcMacroHostPlugin::try_new(instances)?);
+                Ok((component_id, plugin))
+            })
+            .collect::<Result<HashMap<_, _>>>()?;
+
+        for (component_id, suite) in plugins.iter_mut() {
+            if let Some(proc_macro) = proc_macros.get(component_id) {
+                suite.add(ProcMacroHostPlugin::build_plugin_suite(proc_macro.clone()));
+            }
+        }
+
+        Ok(Self {
+            plugins,
+            proc_macros,
+        })
+    }
+}
+
+// NOTE: Since this structure is used to handle JsonRPC requests, its keys have to be serialized to strings.
+// NOTE: It relies on the implicit contract: CompilationUnitComponentId.repr == PackageId.to_serialized_string().
```

Comment:
```suggestion
// INVARIANT: Correct usage of this struct during proc macro server <-> LS communication relies on the implicit contract that keys of `macros_for_packages` are of form `PackageId.to_serialized_string()` which is always equal to `scarb_metadata::CompilationUnitComponentId.repr`.
```

Could be in `///` doc too since it's rly important.

Reply:
Done